zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProvider.java

```diff
@@ -0,0 +1,163 @@
+import org.apache.zookeeper.common.ZKConfig;
+import org.apache.zookeeper.data.Id;
+import org.apache.zookeeper.server.ServerCnxn;
+import org.apache.zookeeper.server.auth.ServerAuthenticationProvider;
+import org.apache.zookeeper.server.auth.X509AuthenticationUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A ServerAuthenticationProvider implementation that does both authentication and authorization for protecting znodes from unauthorized access.
+ * Znodes are grouped into domains according to their ownership, and clients are granted access permission to domains.
+ * Authentication mechanism is same as in X509AuthenticationProvider.
+ * Authorization is done by checking with clients' URI (uniform resource identifier) in ACL metadata for matched domains.
+ * Support client types: single domain client, super user client, client w/o matched domain.
+ * Optional features include: set client Id as znode ACL when creating nodes, add additional public read ACL to znodes based on path
+ */
+public class X509ZNodeGroupAclProvider extends ServerAuthenticationProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(X509ZNodeGroupAclProvider.class);
+  private final String logStrPrefix = this.getClass().getName() + ":: ";
+  private final X509TrustManager trustManager;
+  private final X509KeyManager keyManager;
+  static final String ZOOKEEPER_ZNODEGROUPACL_SUPERUSER = "zookeeper.znodeGroupAcl.superUser";
+
+  public X509ZNodeGroupAclProvider() {
+    ZKConfig config = new ZKConfig();
+    this.keyManager = X509AuthenticationUtil.createKeyManager(config);
+    this.trustManager = X509AuthenticationUtil.createTrustManager(config);
+  }
+
+  public X509ZNodeGroupAclProvider(X509TrustManager trustManager, X509KeyManager keyManager) {
+    this.trustManager = trustManager;
+    this.keyManager = keyManager;
+  }
+
+  @Override
+  public KeeperException.Code handleAuthentication(ServerObjs serverObjs, byte[] authData) {
+    // Get x509 certificate
+    ServerCnxn cnxn = serverObjs.getCnxn();
+    X509Certificate[] certChain = (X509Certificate[]) cnxn.getClientCertificateChain();
+    if (certChain == null || certChain.length == 0) {
+      LOG.error(logStrPrefix + "No x509 certificate is found.");
+      return KeeperException.Code.AUTHFAILED;
+    }
+    X509Certificate clientCert = certChain[0];
+
+    if (trustManager == null) {
+      LOG.error(logStrPrefix + "No trust manager available to authenticate session 0x{}",
+          Long.toHexString(cnxn.getSessionId()));
+      return KeeperException.Code.AUTHFAILED;
+    }
+
+    try {
+      // Authenticate client certificate
+      trustManager.checkClientTrusted(certChain, clientCert.getPublicKey().getAlgorithm());
+    } catch (CertificateException ce) {
+      LOG.error(logStrPrefix + "Failed to trust certificate for session 0x{}",
+          Long.toHexString(cnxn.getSessionId()), ce);
+      return KeeperException.Code.AUTHFAILED;
+    }
```

Comment:
Is there any reason we're keeping this duplicate code?

Reply:
These are some null checks and some variable assignments (for example clientCert) which are going to be used in the code below. So I think there's no point making it a util method to avoid duplicated code.

Reply:
I still think it's possible to factor this out into 2 methods?
1)
`    X509Certificate[] certChain = (X509Certificate[]) cnxn.getClientCertificateChain();
    if (certChain == null || certChain.length == 0) {
      LOG.error(logStrPrefix + "No x509 certificate is found.");
      return KeeperException.Code.AUTHFAILED;
    }
    X509Certificate clientCert = certChain[0];`

2)
`    if (trustManager == null) {
      LOG.error(logStrPrefix + "No trust manager available to authenticate session 0x{}",
          Long.toHexString(cnxn.getSessionId()));
      return KeeperException.Code.AUTHFAILED;
    }

    try {
      // Authenticate client certificate
      trustManager.checkClientTrusted(certChain, clientCert.getPublicKey().getAlgorithm());
    } catch (CertificateException ce) {
      LOG.error(logStrPrefix + "Failed to trust certificate for session 0x{}",
          Long.toHexString(cnxn.getSessionId()), ce);
      return KeeperException.Code.AUTHFAILED;
    }`

Reply:
I extracted it out as one method instead of 2 - in your 2nd method it still needs variable `certainChain` and `clientCert`, which are get in your 1st method. So it will be several lines of redundant duplicated code if we separate it into 2 methods.